Get rid of downcast<>() in ContainerNode::rootNode()

cdumez opened the pull request:
#### be36acd769261df22b97655ff0579f150596ffcc
<pre>
Get rid of downcast&lt;&gt;() in ContainerNode::rootNode()
<a href="https://bugs.webkit.org/show_bug.cgi?id=270144">https://bugs.webkit.org/show_bug.cgi?id=270144</a>

Reviewed by Ryosuke Niwa.

Get rid of downcast&lt;&gt;() in ContainerNode::rootNode(), for performance reasons.

* Source/WebCore/dom/ContainerNode.h:
(WebCore::ContainerNode::rootNode const):
(WebCore::ContainerNode::protectedRootNode const):

Canonical link: <a href="https://commits.webkit.org/275416@main">https://commits.webkit.org/275416@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3056b03f9dc59996d4b0002c0e2af0fa1a53e6d2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/41782 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/20796 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/44164 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/44359 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/37875 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/44089 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/23939 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/18127 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/34532 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/42356 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/17723 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/35982 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/15202 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/15414 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/37006 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/45753 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/37967 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/37328 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/41080 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/16597 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/13630 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/39524 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/18216 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9364 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/18274 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/17860 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/dom/ContainerNode.h:
```diff
@@ -59,8 +59,8 @@ class ContainerNode : public Node {
 
-    ContainerNode& rootNode() const { return downcast<ContainerNode>(Node::rootNode()); }
-    Ref<ContainerNode> protectedRootNode() const { return downcast<ContainerNode>(Node::rootNode()); }
+    ContainerNode& rootNode() const { return uncheckedDowncast<ContainerNode>(Node::rootNode()); }
```
Hm... it's not immediately obvious to me that this cast is always safe.
I get that ContainerNode can't have non-ContainerNode as its parent but maybe it's best to play safe & check that at runtime.

cdumez replied:
Ok, I'll look into making it more obvious. I was trying to avoid it because it duplicates some code.

cdumez replied:
Ok. Hopefully things are clearer now. I still have a tiny regression on Intel when turning on the RELEASE_ASSERT() in downcast<>() so I'm trying to be aggressive enough to turn it on.

rniwa commented on Source/WebCore/dom/ContainerNode.cpp:
```diff
@@ -1244,4 +1244,9 @@ HTMLCollection* ContainerNode::cachedHTMLCollection(CollectionType type)
 
+ContainerNode& ContainerNode::traverseToRootNode() const
+{
+    return traverseToRootNodeGeneric(*this);
```
How about traverseToRootNodeInternal?

cdumez replied:
Fine by me. I'll rename.

cdumez committed:
Get rid of downcast<>() in ContainerNode::rootNode()
https://bugs.webkit.org/show_bug.cgi?id=270144

Reviewed by Ryosuke Niwa.

Get rid of downcast<>() in ContainerNode::rootNode(), for performance reasons.

* Source/WebCore/dom/ContainerNode.h:
(WebCore::ContainerNode::rootNode const):
(WebCore::ContainerNode::protectedRootNode const):

Canonical link: https://commits.webkit.org/275416@main
